pac4j-core/src/main/java/org/pac4j/core/client/IndirectClient2.java

```diff
@@ -0,0 +1,148 @@
+/*
+  Copyright 2012 - 2015 pac4j organization
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+package org.pac4j.core.client;
+
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.authenticator.Authenticator;
+import org.pac4j.core.credentials.Credentials;
+import org.pac4j.core.credentials.authenticator.LocalCachingAuthenticator;
+import org.pac4j.core.credentials.extractor.Extractor;
+import org.pac4j.core.exception.CredentialsException;
+import org.pac4j.core.exception.RequiresHttpAction;
+import org.pac4j.core.exception.TechnicalException;
+import org.pac4j.core.profile.CommonProfile;
+import org.pac4j.core.profile.creator.AuthenticatorProfileCreator;
+import org.pac4j.core.profile.creator.ProfileCreator;
+import org.pac4j.core.redirect.Redirector;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.core.util.InitializableWebObject;
+
+/**
+ * New indirect client type using the {@link Redirector}, {@link Extractor}, {@link Authenticator}
+ * and {@link ProfileCreator} concepts.
+ * 
+ * @author Jerome Leleu
+ * @since 1.9.0
+ */
+public abstract class IndirectClient2<C extends Credentials, U extends CommonProfile> extends IndirectClient<C, U> {
```

Comment:
But that wont matter, since sem ver indicates that you can change Impl API across minor versions. So I'd rather we get rid of the old client, and replace it with this "2" one :)